ESQL Support loading points from source into WKB blocks

craigtaverner opened the pull request:
The work in https://github.com/elastic/elasticsearch/pull/102177 added support for `geo_point` and `cartesian_point` to ES|QL. However, that work relied on ES|QL existing support for doc-values, and the `LongBlock` in particular, to handle all points inside the compute engine. This made the implementation relatively easy, but did come at the cost of displaying points quantized to the encoded grid used in the Lucene point index. For most use cases, this is fine as the error is very, very small. In particular this made it useful for Kibana to start work on displaying `geo_point` data in the Kibana map.

However, for the GA version of spatial support in ES|QL, we definitely want to have the users see exactly the same precision that they ingested the data with. To achieve this we needed to enable reading from source for points. The design is to use WKB encoding within a BytesRefBlock to store all kinds of Geometries, so this will also be of use for `geo_shape` and  `shape` going forward.

The initial work started in https://github.com/elastic/elasticsearch/pull/102880, but here we are replacing the use of `PointBlock` with `BytesRefBlock` and evaluating the difference.

- [x] Support reading from source in the GeoPointFieldMapper and PointFieldMapper `blockLoader` method
- [x] Support new WKB encoder for BytesRefBlock
- [x] Check TransportVersion for plan serialization of literal points (encoded longs in 8.12 and WKB in 8.13)
- [ ] Determine if we need to support TransportVersion for compatibility with 8.12 for the Blocks
- [x] Re-think the use of a common parent class for GeoPoint and CartesianPoint (If we create a `Coords2D` we can revert `SpatialPoint` to an interface)
- [x] Improve support for points with `ComparisonMapper` for equality but not inequality
- [x] Remove interim support of `PointBlock` (inherited from https://github.com/elastic/elasticsearch/pull/102880)
- [ ] Get `GeoPointFieldMapperTests` working for `rowStrideReader`
- [x] The `WKBTopNEncoder` needs to encode in a binary format that does not use `0` or the continuation byte. This rules out WKB, so we use WKT. This is inefficient. Since spatial types don't require sorting, we could switch to length-prefixed WKB.
- [ ] Review all `TODO` entries to see which can/should be fixed in this PR
- [ ] Support planner control over whether we use doc-values (column at a time) or source (row-stride)

craigtaverner committed:
ESQL reading points from source

Update docs/changelog/102880.yaml

Make SpatialPoint a concrete class

Simplifies code in a few places, and is a step towards easier implementation of other features, especially CRS support and ESQL support.

Continued work on point block support

Spotless checks for generated code

Disable test for POINT in a few complex cases

* MultivalueDedupeTests seems to order POINTs incorrectly
* TopNOperatorTests might need to exclude POINT always unless we want to support POINT in TopN results
* BlockHashRendomizedTests has failures on POINT, perhaps an issue with different hashcode calculations?

Added missing generated code for building point types

Get CsvTests working with more support for PointBlock

Fixed ToString and ToLong tests after change of point from encoded

Added missing new generated files

Updated changelog to be ESQL specific.

A hint as to where to tie into the planner for doc-values vs source

Small refinement to PointArray/Vector memory estimates

Small refinement to PointArray/Vector memory estimates

Small refinement to PointArray/Vector memory estimates

Reorganize BlockFectory to improve symmetry between Double and Point

Support read/write Point in StreamInput/Output

Fix some more tests

Fixed asymmetry between Geo and Cartesian field types

This was expressed as test failures for CartesianPoint since it was still reading from doc-values instead of from source, while the rest of the stack expected source.

Disable failing field type tests for now

Currently reading from source for GeoPointFieldMapper is not properly supported in the tests

Fixed failing tests with unsupported types and precision change

Fix tests that still used longs for spatial points

Added source values to spatial types in textFormatting tests

Fixed EsqlQueryResponse to map point back to geo_point after deserialization

Fixed binary comparison tests for spatial types

In particular, we do not support inequality operators on spatial types.
For example: point < point, is not a meaningful expression.

Fixed mixed-cluster tests for csv-tests and spatial precision

Fixed mixed-cluster tests for esql types and point precision

More tests and use warning for parse error

Remove SpatialPoint from StreamInput/Output

Added fromStats() for spatial block values from doc-values

Serialize GeoPoint and SpatialPoint in ESQL plan

Retain point types in plan serialization

Fixup after rebase

* Removed support for unsigned_long from ToGeoPoint and ToCartesianPoint, since the mapping from unsigned_long to long allows for invalid values
* Support new X-BigArrayBlock introduced by Nhat

Fix test where there can be duplicated hashcodes

When cartesian decoding fails, we should test and mimic that failure

craigtaverner committed:
Initial support for BytesRefBlock to support WKB for Geometries

This commit does not yet remove support for PointBlock, since we are investigating how to kabe multiple block types backing the same geometries.

craigtaverner committed:
Update docs/changelog/103698.yaml

elasticsearchmachine commented:
Hi @craigtaverner, I've created a changelog YAML for you.

craigtaverner committed:
Fixed changelog files

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/BlockSourceReader.java:
```diff
@@ -149,6 +192,75 @@ public String toString() {
+                    Geometry geometry = WellKnownText.fromWKT(GeometryValidator.NOOP, false, wkt);
+                    if (geometry instanceof Point point) {
+                        // TODO: perhaps we should not create points for later GC here, and pass in primitives only?
+                        ((BlockLoader.PointBuilder) builder).appendPoint(new SpatialPoint(point.getX(), point.getY()));
```
Indeed, we should try not to create any more garbage than we have to here. The `_source` loading is going to allocate - it has to do so to get at `_source`, but we should avoid it when we can.

craigtaverner replied:
We've simplified this now, by generating WKB in the value fetcher and not converting types.

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/BlockSourceReader.java:
```diff
@@ -149,6 +192,75 @@ public String toString() {
+                    } else {
+                        throw new IllegalArgumentException("Cannot convert geometry into point:: " + geometry.type());
+                    }
+                } catch (Exception e) {
```
Probably best to catch precise exceptions here.

craigtaverner replied:
Done. It was two different exceptions, so I'll use `catch (IOException | ParseException e)`

nik9000 commented on x-pack/plugin/esql/compute/build.gradle:
```diff
@@ -30,27 +30,30 @@ spotless {
+  var longProperties     = prop("Long", "long", "LONG", "Long.BYTES", "LongArray")
+  var doubleProperties   = prop("Double", "double", "DOUBLE", "Double.BYTES", "DoubleArray")
+  var bytesRefProperties = prop("BytesRef", "BytesRef", "BYTES_REF", "org.apache.lucene.util.RamUsageEstimator.NUM_BYTES_OBJECT_REF", "")
+  var pointProperties    = prop("Point", "SpatialPoint", "POINT", "16", "ObjectArray<SpatialPoint>")
```
Ah! Yeah, that probably needs to be `double[len * 2]` or something.

craigtaverner replied:
Deleted the entire PointBlock, so this is removed.

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/topn/WKBTopNEncoder.java:
```diff
@@ -0,0 +1,64 @@
+ * Since WKB can contain bytes with zero value, which are used as terminator bytes, we need to encode differently.
+ * Our initial implementation is to re-write to WKT and encode with UTF8TopNEncoder.
+ * This is likely very inefficient.
+ * We cannot use the UTF8TopNEncoder as is, because it removes the continuation byte, which could be a valid value in WKB.
```
Ooof. Ouch.

craigtaverner replied:
I was thinking a nicer option would be to base64 encode the binary. Might be more space efficient than converting WKB to WKT and back.

iverase replied:
That's terrible that we cannot support generic binary data because we are giving special meaning so some bytes.

iverase replied:
we should runlen prefix rows and columns. It is more efficient for parsing as we don't need to check every byte to be a special one and it provides support for generic binary data. The overhead it adds in size should be pretty small.

iverase replied:
ok, not sure what you want to use the utf8 encoder when we are not utf8. 

 I think geometries are not sortable so this can be implemented as non sortable and we can encode it / decode it using a runlen prefix (saving how many bytes are to be written / we need to read ).

nik9000 replied:
++. We can just runlen prefix it.

If we need to sort the geometries one day we'd come up with an encoding.

nik9000 replied:
We probably should just call this `UnsortableViableLengthBytesEncoder` or something.

nik9000 commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/util/SpatialCoordinateTypes.java:
```diff
@@ -116,4 +91,51 @@ public SpatialPoint stringAsPoint(String string) {
+    public long wkbAsLong(BytesRef wkb) {
+        Geometry geometry = WellKnownBinary.fromWKB(GeometryValidator.NOOP, false, wkb.bytes, wkb.offset, wkb.length);
+        if (geometry instanceof Point point) {
+            return pointAsLong(point.getX(), point.getY());
```
This is all pretty allocate-y, but it's the tools we have.

nik9000 commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/util/SpatialCoordinateTypes.java:
```diff
@@ -9,14 +9,16 @@
 import org.elasticsearch.geometry.Geometry;
 import org.elasticsearch.geometry.Point;
 import org.elasticsearch.geometry.utils.GeometryValidator;
+import org.elasticsearch.geometry.utils.WellKnownBinary;
```
I think this enum here is kind of at the core of the implementation - we "read" the wkb using this. I'd prefer to have something that can manage to allocate less, but that can come with time.

craigtaverner replied:
Yeah, this enum is collecting some cruft as we refactor the code, but I think we'll have the opportunity to cleanup q bit here once we have geo_shape completed and know the full scope of what is really needed here. The main pain is with the docs-values using such different encoding between the four main types... that is such an unnecessary hassle! But we cannot escape it because Lucene enforces that.

craigtaverner committed:
Remove support for PointBlock

We will use only WKB encoded in BytesRefBlock for now

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_points_from_source_wkb

craigtaverner committed:
One change from code review

craigtaverner commented on server/src/main/java/org/elasticsearch/index/mapper/AbstractGeometryFieldMapper.java:
```diff
@@ -136,6 +139,23 @@ protected Object parseSourceValue(Object value) {
+        @Override
+        public BlockLoader blockLoader(BlockLoaderContext blContext) {
+            // TODO: If we have doc-values we have to use them, due to BlockSourceReader.columnAtATimeReader() returning null
+            if (blContext.forStats() && hasDocValues()) {
```
There are many options for this. I was looking at it from two perspectives:

* From the field type, this relates to whether or not to use doc-values, but considering how most types favor doc-values and the actual doc-values and source values are identical, this is very specific to spatial types, so seems like the wrong perspective.
* From the planner side, what is the intended use of this field, with there being two somewhat different uses: just pass through and include in the results, or consume as part of an aggregation. You originally suggested `forDisplay`, which is the complement of my preference for `forStats`.

I feel like `transformed` is an obsfucation of `forStats` as it does not directly describe the intended use (transformed by stats?), nor does it describe the consequences exactly (transformed into doc-values?).

Are you suggesting that `transformed` might be a widening? Are there other cases where we might transform the value in such a way that we would like to use `doc-values` in the field type? Perhaps `eval` with functions that transform the value? I can imagine such cases, like `EVAL proximity = ST_BUFFER(road, 10, 'join=mitre mitre_limit=1.0')` which produces a new geometry and as such can get a performance benefit by not reading from source? That is definitely worth thinking about!

craigtaverner replied:
Lots of interesting ideas here, but I think we should leave this for the next PR where we bring back doc-values in the planner. For now a simple boolean on/off flag is sufficient for me to get the tests unmuted.

craigtaverner committed:
Removed some leftovers from the PointBlock removal

craigtaverner committed:
Some code-review and TODO checks

craigtaverner committed:
Some code-review and TODO checks, and deal with Plan serialization

For plan serialization we needed to care that 8.12 expected point literals to serialize as encoded long values.

craigtaverner committed:
Fixed failing tests with point WKT rendering

craigtaverner committed:
Revert stab at implementing forStats for doc-values vs source

We can try this again in another PR

craigtaverner committed:
Disabled failing test

craigtaverner committed:
Use max precision when serializing points to XContent

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

elasticsearchmachine commented:
Pinging @elastic/es-ql (Team:QL)

elasticsearchmachine commented:
Pinging @elastic/elasticsearch-esql (:Query Languages/ES|QL)

craigtaverner committed:
Removed unused class from PointBlock code

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_points_from_source_wkb

breskeby reviewed (approved):
build script change looks ok

craigtaverner committed:
Simplifications from code review

craigtaverner commented on server/src/main/java/org/elasticsearch/index/mapper/AbstractGeometryFieldMapper.java:
```diff
@@ -136,6 +139,23 @@ protected Object parseSourceValue(Object value) {
+
+        @Override
+        public BlockLoader blockLoader(BlockLoaderContext blContext) {
+            // TODO: If we have doc-values we have to use them, due to BlockSourceReader.columnAtATimeReader() returning null
```
I removed everything but the read-from-source mode now. I'll bring back the doc-values in the next PR.

iverase commented on server/src/main/java/org/elasticsearch/index/mapper/AbstractGeometryFieldMapper.java:
```diff
@@ -136,6 +139,14 @@ protected Object parseSourceValue(Object value) {
+        public BlockLoader blockLoader(BlockLoaderContext blContext) {
+            // Currently we can only load from source in ESQL
+            return new BlockSourceReader.GeometriesBlockLoader(
+                valueFetcher(blContext.sourcePaths(name()), nullValue, GeometryFormatterFactory.WKT)
```
In https://github.com/elastic/elasticsearch/pull/103461, we added support to generate WKB. I think we should use that here so we don't need to transform the WKT into WKB later, it is wasteful.

craigtaverner replied:
Good point. I see the option to generate WKB directly in the value fetcher, so we don't go via WKT, but directly to WKB. I'm busy working on GeoPointFieldMapperTests at the moment, so I'll investigate this too.

craigtaverner replied:
Done

iverase commented on server/src/main/java/org/elasticsearch/index/mapper/BlockSourceReader.java:
```diff
@@ -149,6 +175,41 @@ public String toString() {
+        @Override
+        protected void append(BlockLoader.Builder builder, Object v) {
+            if (v instanceof SpatialPoint point) {
+                BytesRef wkb = new BytesRef(WellKnownBinary.toWKB(new Point(point.getX(), point.getY()), ByteOrder.LITTLE_ENDIAN));
```
I am very confused here, do we really generate points now? Do we really need this? the idea is that we only load WKB?

craigtaverner replied:
I think this was defensive coding, for the variety of possible loading we will see going forward. The source value fetcher was generating WKT (but I can change that to WKB), doc-values was generating longs, and stored fields would generate points themselves (I assumed, but have not written support for that yet).

iverase replied:
Lets remove defensive coding, I think we should only accept WKB here.

nik9000 replied:
++ on keeping this list as small as possible. Accepting `Object` for `append` makes this whole difficult to understand, but that's how reading `_source` works at the moment and we ain't changing it here. So folks reading this have the easiest time I think we should only handle the actual things we expect to get. Is that *just* `BytesRef`, right?

craigtaverner replied:
Removed support for SpatialPoint

iverase commented on server/src/main/java/org/elasticsearch/index/mapper/BlockSourceReader.java:
```diff
@@ -149,6 +175,41 @@ public String toString() {
+                BytesRef wkb = new BytesRef(WellKnownBinary.toWKB(new Point(point.getX(), point.getY()), ByteOrder.LITTLE_ENDIAN));
+                ((BlockLoader.BytesRefBuilder) builder).appendBytesRef(wkb);
+            } else if (v instanceof String wkt) {
+                try {
```
We can generate WKB from the loaders so we don't need to do this, seems wasterful.

craigtaverner replied:
Agreed, I'll look into changing this now.

craigtaverner replied:
Done

iverase commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/ElementType.java:
```diff
@@ -61,6 +63,10 @@ public static ElementType fromJava(Class<?> type) {
             elementType = DOUBLE;
         } else if (type == String.class || type == BytesRef.class) {
             elementType = BYTES_REF;
+        } else if (SpatialPoint.class.isAssignableFrom(type)) {
```
This is very obscure to me but I read it like one type is mapped to one and only one type. This seems problematic with geo as depending how the data is consumed we might want to generate a different element type? I don't see how this is helping us on the generic solution and makes me very worried,

craigtaverner replied:
I still think that when we re-enable doc values (next PR, I assume) this will be solvable. I've seen issues like this before, when I needed to pass down additional information to differentiate between types, and it was generally possible to do so, so I assume the same will be true here.

nik9000 replied:
I'm not a huge fan of `fromJava` in general. I think it's mostly useful when we parse constants and then want to convert them into types. And for tests. I suppose it's really useful in tests to be like "make me a block from these geometries", but, like you say, there are actually two choices of kinds of block. At least there are for points. Oh boy.

iverase replied:
Ir un the testa and this code seems never to be exercised. Please either add test to exercise the code or remove it. I think we should never have those classes here so I am more inclined to remove them.

craigtaverner replied:
Removed support for SpatialType and Geometry. Turns out this was only excercised by tests that should themselves have been generating WKB, so I fixed the tests, and removed this code.

iverase commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/PlannerUtils.java:
```diff
@@ -232,11 +232,12 @@ public static ElementType toElementType(DataType dataType) {
         if (dataType == EsQueryExec.DOC_DATA_TYPE) {
             return ElementType.DOC;
         }
+        // TODO: Spatial types can be read from source into BYTES_REF, or read from doc-values into LONG
```
As above, this is another place wherre we seem to imply that one field type can only generate one Element type, is this a blocker to return different data structures depnding on how the data is going to be consumed?

craigtaverner replied:
Actually, this case is an easier one. We're planning to make different data types for the doc-values version, so there will be no ambiguity in this case. The planner will plan the type it needs, either the source type or the doc values type, and then methods like this will simply see these as two different types. This is at least the idea I've discussed with Nik, and I think it makes sense.

nik9000 replied:
Indeed, this is another place where we were assuming all DataTypes go to a single element type. I don't know the right way to handle this. Some kind of information in `dataType` or something. I dunno.

I'm quite ok waiting on solving this for another PR though. We *know* this is going to be a big part of the next PR.

iverase commented on server/src/main/java/org/elasticsearch/index/mapper/BlockSourceReader.java:
```diff
@@ -131,6 +139,24 @@ public RowStrideReader rowStrideReader(LeafReaderContext context) {
         }
     }
 
+    public static class GeometriesBlockLoader extends SourceBlockLoader {
```
It makes me feel uneasy, cannot we reuse the BytesRef  reader as we are reading an array of bytes?

craigtaverner replied:
Right now we need this to get to the code in the 'Geometries' BlockSourceReader, but as that can get simplified, perhaps this can too.

iverase reviewed (commented):
I had a look into the PR. I cannot really review it as I lack knowledge of how the compute engine works so I can only comment on my expectations and what is missing from this expectations.

The idea os this PR is that provides the capacity for reading a point from _source and provide to the user as WKT on the result of a query. For that I expect the transport format to be WKB and only generate the WKT when projecting the value to the user. 
Therefore I don't understand all those transformations between WKT to WKB and back, Note that for big geometries this can be very costly.  

More over, we event transform from WKB to spatial point, do we really need that abstractions here? if so why? 

I still feel that the engine only allows one field type to generate one data structure and I don't see in this PR hoe we can go around that limitation. I started wondering if we need to change how we handle geo data in Elasticsearch in order to be able to use the compute engine. The columnar approach taken here is adding some constraints that did not exist before.

craigtaverner committed:
Remove intermediate WKT from GeoPointFieldMapper

Previously we were reading source, mapping to WKT and then later mapping to WKB. Now just map directly to WKB.

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/BlockLoader.java:
```diff
@@ -348,6 +349,11 @@ interface BlockFactory {
+        /**
+         * Build a builder to load {@link SpatialPoint}s backed by WKB in BytesRefBlock.
+         */
+        BytesRefBuilder geometries(int expectedCount);
```
Could we remove this and use `bytesRefs`? If so, maybe we could reuse `BlockSourceLoader.BytesRefsBlockLoader`  like Ignacio was saying. It seems more likely.

craigtaverner replied:
OK, I've removed that now, but could not remove GeometriesBlockLoader because it uses Geometries which does:

```
((BlockLoader.BytesRefBuilder) builder).appendBytesRef(new BytesRef(wkb));
```

While the equivalent one in the BytesRefsBlockLoader does:

```
((BlockLoader.BytesRefBuilder) builder).appendBytesRef(toBytesRef(scratch, (String) v));
```

So the BytesRefs one assuming incoming data is a String object, while the Geometries one assumes it is a `byte[]`. The difference seems quite important to me.

nik9000 commented on test/framework/src/main/java/org/elasticsearch/index/mapper/MapperTestCase.java:
```diff
@@ -1243,7 +1243,7 @@ public final void testBlockLoaderFromColumnReader() throws IOException {
     }
 
-    public final void testBlockLoaderFromRowStrideReader() throws IOException {
+    public void testBlockLoaderFromRowStrideReader() throws IOException {
```
ah! you removed `final` so you could silence it. Probably worth a comment if we're going to commit this. `final` is nice here.

craigtaverner replied:
Added comment

nik9000 commented on test/framework/src/main/java/org/elasticsearch/index/mapper/TestBlock.java:
```diff
@@ -57,6 +57,18 @@ public BytesRefsBuilder appendBytesRef(BytesRef value) {
+                class GeometriesBuilder extends TestBlock.Builder implements BlockLoader.BytesRefBuilder {
+                    @Override
+                    public GeometriesBuilder appendBytesRef(BytesRef value) {
+                        add(BytesRef.deepCopyOf(value));
```
This *looks* like we can just use `bytesRefs` above.

I think we needed this when we had a totally different block type - one that natively supported points. But we don't so much now.

craigtaverner replied:
Removed now, along with the simplifications above

nik9000 commented on test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java:
```diff
@@ -1204,17 +1206,7 @@ public static SpatialPoint randomGeoPoint() {
-                return y;
-            }
-        };
+        return new SpatialPoint(x, y);
```
Oh this is nicer!

craigtaverner replied:
This is unfortunately reverted to keep SpatialPoint as an interface. Getting is back to a concrete class will require more work, as the current attempt was seen as an awkward compromise.

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/BlockUtils.java:
```diff
@@ -207,13 +212,23 @@ public static void appendValue(Block.Builder builder, Object val, ElementType ty
         }
     }
 
+    private static BytesRef spatialToBytesRef(Object val) {
```
Probably worth a rename and a javadoc. I'd love the name to imply that it also converts strings to bytes refs.

nik9000 replied:
Not sure a good name. `allJavaTypesToBytesRef`. Or, even, `toBytesRef` is an ok name, I guess. But I'm just bike shedding.

craigtaverner replied:
I felt like `toBytesRef` would clash with the previous existing method, so I wanted a new name and `toBytesRefAlsoConsideringSpatial` seemed even worse.

craigtaverner replied:
I went for `toBytesRef`, we can all follow the stack anyway.

craigtaverner replied:
This is now entirely removed, as it was only really needed for tests that themselves were not correctly written.

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/X-BigArrayBlock.java.st:
```diff
@@ -8,14 +8,14 @@
 
 import org.apache.lucene.util.RamUsageEstimator;
-import org.elasticsearch.common.util.$if(boolean)$Bit$else$$Type$$endif$Array;
+import org.elasticsearch.common.util.$Array$;
```
:+1:

craigtaverner replied:
I was expecting someone to ask me to remove this, so I'm glad you like it. When I had PointBlock, this was very necessary, but even without PointBlock I think this looks a lot nicer, so pleased to keep it.

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/X-BigArrayVector.java.st:
```diff
@@ -55,19 +55,22 @@ public final class $Type$BigArrayVector extends AbstractVector implements $Type$
             filtered.set(i, values.get(positions[i]));
         }
-   $endif$
+    $endif$
```
Thanks.

craigtaverner replied:
Instead of just reverting all the generated code stuff I did, I removed all the PointBlock stuff, and then re-reviewed what was left and kept what I thought still seemed like an improvement, even if not necessary for this PR.

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/X-MultivalueDedupe.java.st:
```diff
@@ -50,9 +46,7 @@ $if(BytesRef)$
-    private static final int ALWAYS_COPY_MISSING = 300;
-$elseif(long)$
+$else$
     private static final int ALWAYS_COPY_MISSING = 300;
```
I'd intended to make this something you'd have to fill out. Maybe I should have had an `else` section that says "you have to benchmark this. With actual instructions on how. Like, there's a benchmark in `MultivalueDedupeBenchmark` but I don't 100% remember how I did the benchmarking.

craigtaverner replied:
It is likely long and int would behave all that differently?

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/topn/TopNEncoder.java:
```diff
@@ -40,6 +40,10 @@ public interface TopNEncoder {
      */
     VersionTopNEncoder VERSION = new VersionTopNEncoder();
+    /**
+     * An encoder for UTF-8 text.
```
s/UTF-8 text/well know binary representations of geometries/.

craigtaverner replied:
Well, I've completely deleted the WKB encoder now!

nik9000 commented on x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BasicBlockTests.java:
```diff
@@ -369,12 +373,12 @@ public void testConstantDoubleBlock() {
     }
 
-    public void testBytesRefBlock() {
+    private void testBytesRefBlock(Supplier<BytesRef> byteArraySupplier, boolean chomp, org.mockito.ThrowingConsumer<BytesRef> assertions) {
```
`CheckedConsumer`? I'm not used to seeing us reference the guts of mockito.

craigtaverner replied:
Well, the difference between the existing UTF8 encoder test and the new WKB was tiny, and only needed this one extra assertion. This seemed the cleanest way to achieve that.

nik9000 commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java:
```diff
@@ -906,19 +904,12 @@ private static String typeErrorMessage(boolean includeOrdinal, List<Set<DataType
-            ),
-            "cartesian_point or long or string or unsigned_long"
+            Set.of(EsqlDataTypes.CARTESIAN_POINT, DataTypes.KEYWORD, DataTypes.LONG, DataTypes.TEXT, DataTypes.NULL),
+            "cartesian_point or long or string"
```
Sorry this code is such a mess. It's trying to be cute and guess error messages but I think it can get a bit annoying to maintain.

nik9000 reviewed (commented):
I'm happy with it. It's not the end, very much not. But I think it's good progress and we can get it in.

craigtaverner committed:
Refactor WKBTopNEncoder to use length prefix encoding

The previous approach of using WKT was expensive and wasteful. Since the types are unsortable, we can just encode them as is.

craigtaverner committed:
Fix failing test

craigtaverner committed:
Remove WKBTopNEncoder since DefaultUnsortableTopNEncoder can do the job

We move the support for BytesRef encoding to DefaultUnsortableTopNEncoder.

craigtaverner committed:
Cleanup from code review

iverase commented on server/src/main/java/org/elasticsearch/index/mapper/AbstractGeometryFieldMapper.java:
```diff
@@ -136,6 +139,14 @@ protected Object parseSourceValue(Object value) {
         }
+
+        @Override
+        public BlockLoader blockLoader(BlockLoaderContext blContext) {
```
Can we move this to AbstractPointGeometryFieldMapper and remove the override method in AbstractShapeGeometryFieldMapper?

craigtaverner replied:
There is no intermediate class common between Geo and Cartesian points, so that would require quite a bit of restructuring. Besides, the plan is to support geo_shape very soon, so we would just move back to this then anyway.

iverase replied:
I thought `AbstractPointGeometryFieldMapper` is common for geo and cartesian points. I know it will change but we should keep PRs to the minimum please or it makes reviewing very difficult.

nik9000 reviewed (approved):
I'm good with it if @iverase is.

iverase reviewed (changes requested):
I have been trying to justify making SpatialPoint a concrete class and I cannot, a point is ether geo or cartesian and we should never allow create spatial points by itself. Therefore I need to ask to revert the change in SpatialPoint to be a concrete class, it is just wrong.

iverase commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/BlockUtils.java:
```diff
@@ -214,6 +219,16 @@ public static void appendValue(Block.Builder builder, Object val, ElementType ty
+        }
+        if (val instanceof Geometry geometry) {
+            return new BytesRef(WellKnownBinary.toWKB(geometry, ByteOrder.LITTLE_ENDIAN));
+        }
```
I just run the tests and these lines above seem to never be called, can we remove it? Otherwise could you point me where this code is exercise?

iverase replied:
Note that I think we should never have those classes at this point, so I think this is not necessary.

craigtaverner replied:
This is now removed

craigtaverner committed:
Remove concrete SpatialPoint class

SpatialPoint reverts to an interface

craigtaverner committed:
Bring back SpatialPointTests

Even without a concrete SpatialPoint class, this test suite has some utility.
It asserts, for example, that GeoPoint and other implementations of SpatialPoint do not have the same hash-code even if they contain the same coordinates. Likewise the equals method is tested. These tests are valid regardless of whether GeoPoint and CartesianPoint share a common parent class, or merely implement an interface.

craigtaverner committed:
Reduce object creation in some tests

iverase commented on server/src/test/java/org/elasticsearch/common/geo/SpatialPointTests.java:
```diff
@@ -0,0 +1,103 @@
+
+public class SpatialPointTests extends ESTestCase {
+
+    public void testEqualsAndHashcode() {
```
Do we need this class now that we reverted SpatialPoint as concrete class?

iverase replied:
forget it, its fine,

craigtaverner replied:
Initially I had removed it, but then realized it still had value, even with GeoPoint and CartesianPoint not sharing a common parent class, so I brought it back.

iverase commented on server/src/test/java/org/elasticsearch/common/geo/SpatialPointTests.java:
```diff
@@ -0,0 +1,103 @@
+        }
+    }
+
+    public void testCompareTo() {
```
maybe we should add a test where two different subclasses compare to false?

iverase commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/util/SpatialCoordinateTypes.java:
```diff
@@ -61,7 +61,42 @@ public long pointAsLong(double x, double y) {
+            return makePoint(point.getX(), point.getY());
+        }
+
+        private SpatialPoint makePoint(double x, double y) {
```
I don't think we should do this. We should know if we are dealing with geo or cartesian point.

craigtaverner committed:
Removed unused code

iverase commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/BlockUtils.java:
```diff
@@ -214,6 +218,13 @@ public static void appendValue(Block.Builder builder, Object val, ElementType ty
     }
 
+    private static BytesRef toBytesRef(Object val) {
+        if (val instanceof SpatialPoint point) {
```
This is not use either

craigtaverner replied:
removed

craigtaverner committed:
Removed unused code from earlier versions

Earlier versions of point support needed this, but it is no longer needed.

craigtaverner committed:
Tests were testing cases that no longer apply

We moved internal types to WKB, so testing points themselves no longer makes sense.

craigtaverner committed:
Tests were testing cases that no longer apply

We moved internal types to WKB, so testing points themselves no longer makes sense.

craigtaverner committed:
We need to consider WKB in PlanNamedTypes mixed-cluster

craigtaverner committed:
Removed SpatialPoint from PlanNamedTypes

Only tests were passing spatial points in here, and those have been fixed to generate WKB, which is what is used in production code.

craigtaverner committed:
Removed one more artifact of using SpatialPoint in tests

craigtaverner commented on x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/Types.java:
```diff
@@ -124,6 +124,9 @@ public class Types {
 
     static final ClassName BYTES_REF = ClassName.get("org.apache.lucene.util", "BytesRef");
 
+    static final ClassName POINT = ClassName.get("org.elasticsearch.common.geo", "SpatialPoint");
```
This is removed

craigtaverner committed:
Some work towards removing creating SpatialPoints

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_points_from_source_wkb

craigtaverner committed:
Updated GeoPointFieldMapperTests after merge from main

craigtaverner committed:
Get row-stride-reader test working

Based on Niks support for not using synthetic source, but also by using WKT instead of WKB for test assertions. This is partly because it is easier to debug, but also because the test code uses base64 for encoding expected values, and the production code does not. Switching to WKT avoids that test code pitfall.

craigtaverner committed:
Geo flaky tests to work more reliably

craigtaverner committed:
Removed defensive coding in plan serialization

iverase commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanNamedTypes.java:
```diff
@@ -1572,15 +1579,77 @@ static void writeAlias(PlanStreamOutput out, Alias alias) throws IOException {
+                if (value instanceof List<?> list) {
+                    return list.stream().map(v -> mapFromLiteralValue(out, dataType, v)).toList();
+                }
+                if (value instanceof BytesRef wkb) {
```
We should throw an exception if the value is not a BytesRef? I would either cast it directly so we throw a CastClassException or add an else statement? I think I prefer the first one as the error will be clear.

craigtaverner replied:
Done.

craigtaverner committed:
Simplify and ensure error is thrown on wrong type

iverase commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanNamedTypes.java:
```diff
@@ -1572,15 +1579,77 @@ static void writeAlias(PlanStreamOutput out, Alias alias) throws IOException {
+     * This only makes sense during the pre-GA version of ESQL. When we get near GA we want TransportVersion support.
+     * TODO: Implement TransportVersion checks before GA (eg. by adding to StreamInput/StreamOutput directly)
+     */
+    private static Object mapToLiteralValue(DataType dataType, Object value) {
```
and this?

iverase replied:
we need to check the version here as well sorry

craigtaverner committed:
Do version checks on reading plan from PlanStreamInput

iverase reviewed (approved):
LGTM

I think there is work to do on how the response is built but can be done in a follow up PR.

iverase commented on x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/40_unsupported_types.yml:
```diff
@@ -391,6 +391,38 @@ unsupported with sort:
+spatial types with different precision in 8.12.x:
+  - skip:
+      version: " - 8.11.99, 8.13.0 - "
+      reason: "Elasticsearch 8.12 supported geo_point and cartesian_point with doc-values only and precision differences"
```
should be `version: " - 8.11.99, 8.12.99 - "`?

craigtaverner replied:
I believe the range is inclusive, not exclusive, so it is correct as it is. We skip everything earlier than 8.12.0 and everything from 8.13.0 and later. Basically this is the same as saying "test only 8.12.*".

However, looking at https://github.com/elastic/elasticsearch/pull/103947, it seems there is interest in removing all tests that try to test versions other than the current one. I suspect we might be asked to remove this test entirely.

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_points_from_source_wkb

craigtaverner committed:
Mute failing test

We had this muted before, then fixed it, and now muted it again after https://github.com/elastic/elasticsearch/pull/103632 (ESQL: Check field exists before load from _source)

craigtaverner committed:
Use specific TransportVersions for point literal in query plans

Since we might start testing ESQL in serverless any day now, this is a better option, as the previous approach was more stateful (relied on 8.12.x vs. 8.13.x).

craigtaverner committed:
Fixed failing test after merge with main

The code in main was using an optimization that is not supported by GeoPointFieldMapper, but could be done as a future improvement.

craigtaverner committed:
Fixed failing test after merge with main, with nullValue from source

The nullValue is injected into source fetching in such a way that is is epxected to be in source format. So we need to convert it back to a source format. We picked WKT because that is clear and simple in the debugger, but GeoJSON and an object map of lat/long (for geoPoint only) worked too. Curiously a double[]{x,y} did not work, even though it is a valid source format. We did not investigate why.

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_points_from_source_wkb

craigtaverner committed:
Fixed compile error from previous fix with nullValues

craigtaverner committed:
It seems mixed cluster yaml tests don't work

See https://github.com/elastic/elasticsearch/pull/103947

iverase commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into esql_points_from_source_wkb

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_points_from_source_wkb

nik9000 commented:
![party](https://github.com/elastic/elasticsearch/assets/215970/f16c3114-1bfb-480b-aab5-a4d87e96af78)
